iothub/device/src/AmqpTokenRefresher.cs

```diff
@@ -10,68 +10,62 @@ namespace Microsoft.Azure.Devices.Client
+            this.amqpClientSession = amqpClientSession ?? throw new ArgumentNullException("amqpClientSession");
             this.connectionString = connectionString;
             this.audience = audience;
             this.cancellationTokenSource = new CancellationTokenSource();
         }
 
-        public void Cancel()
+        internal void Cancel()
         {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(AmqpTokenRefresher)}.{nameof(Cancel)}");
+
             if (!this.taskCancelled)
             {
                 this.taskCancelled = true;
                 this.cancellationTokenSource.Cancel();
             }
         }
 
-        public async Task SendCbsTokenAsync(TimeSpan timeout)
+        internal async Task RefreshTokenAsync(TimeSpan timeout)
         {
             try
             {
-                if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(IotHubTokenRefresher)}.{nameof(SendCbsTokenAsync)}");
+                if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(AmqpTokenRefresher)}.{nameof(RefreshTokenAsync)}");
 
                 // Send a Cbs Token right away and fork off a task to periodically renew it
-                var cbsLink = this.amqpSession.Connection.Extensions.Find<AmqpCbsLink>();
-
-                // This can throw PutToken failure in error cases
-                var expiresAtUtc = await cbsLink.SendTokenAsync(
-                    this.connectionString,
-                    this.connectionString.AmqpEndpoint,
-                    this.audience,
-                    this.connectionString.AmqpEndpoint.AbsoluteUri,
-                    AccessRightsStringArray,
-                    timeout).ConfigureAwait(false);
-
-                this.SendCbsTokenLoopAsync(expiresAtUtc, timeout);
+                var expiresAtUtc = await this.amqpClientSession.AuthenticateCbs(timeout).ConfigureAwait(false);
+
+                this.RefreshTokenLoopAsync(expiresAtUtc, timeout).ConfigureAwait(false);
             }
             finally
             {
-                if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(IotHubTokenRefresher)}.{nameof(SendCbsTokenAsync)}");
+                if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(AmqpTokenRefresher)}.{nameof(RefreshTokenAsync)}");
             }
         }
 
-        private async Task SendCbsTokenLoopAsync(DateTime expiryTimeUtc, TimeSpan timeout)
+        private async Task RefreshTokenLoopAsync(DateTime expiryTimeUtc, TimeSpan timeout)
         {
             try
             {
-                if (Logging.IsEnabled) Logging.Enter(this, expiryTimeUtc, timeout, $"{nameof(IotHubTokenRefresher)}.{nameof(SendCbsTokenLoopAsync)}");
+                if (Logging.IsEnabled) Logging.Enter(this, expiryTimeUtc, timeout, $"{nameof(AmqpTokenRefresher)}.{nameof(RefreshTokenLoopAsync)}");
 
                 bool continueSendingTokens = await WaitUntilNextTokenSendTime(
```

Comment:
Seems refresh token is ignored if the tkoken is already expired. Not sure if that's the design.

Reply:
It is by design - the WaitUntilNextTokenSendTime will calculate the next refresh time with a buffer. (this code is not changed from the current implementation)

---
In reply to: [251669153](https://github.com/Azure/azure-iot-sdk-csharp/pull/770#discussion_r251669153) [](ancestors = 251669153)